### ISSUE TITLE
Gunicorn ore pure uvicorn in a kubernetes cluster

### ISSUE THREAD

**Issue by mauermbq** [tiangolo/fastapi#1484]
Hi,

I'm wondering if running pure uvicorn in production when using container orchestrators such as Kubernetes is recommended?
Gunicorn is good for scaling uvicorn workers on multi processor environmnets. However with orchestrators there are inbuilt options to automatically autoscale containers up and down depending on CPU and memory usage, thus minimizing the resource usage.
So why running gunicorn in a container which is ideally designed as running process inside only?

**Comment by tiangolo** [comment]
Sure, if you have another way of declaring and controlling how many containers to spin up, maybe in some automated way, and have a load balancer on top to direct requests to them, then it probably makes sense to do that and have just Uvicorn directly.

**Comment by mauermbq** [comment]
>if you have another way of declaring and controlling how many containers to spin up, maybe in some automated way, and have a load balancer on top to direct requests to them

yes, these capabilities are naturally inbuilt in a cloud based kubernetes stack. Anyway a good choice for the deployment of REST based API's

**Comment by Gatsby-Lee** [comment]
For any one who gets here for the same topic, let me put my thought. ( I am still doing researching this topic )

First, we need to understand these first.
* **Uvicorn:** ASGI Server
* **gunicorn:** WSGI HTTP Server

To use Async functionality in FastAPI, we have to use a ASGI Server. In this case, it is uvicorn.
Gunicorn doesn't replace Uvicorn. Gunicorn can be used to manage Uvicorn workers rather than Uvicorn does it.

1. If you run only one FastAPI app in a container in a Pod, Uvicorn might be enough because anyway you will have one Uvicorn worker that runs FastAPI app. For this case, I think you want to handle traffic by bringing up more Pods

2. If you run more than one FastAPI app, it means you have options for worker management ( or worker process management  )
  2.1. using Gunicorn to launch / manage Uvicorn workers ( gunicorn main:app --workers=2 -k uvicorn.workers.UvicornWorker )
      FYI, ( "--threads" option in Gunicorn with UvicornWorker doesn't affect. )
  2.2. using Uvicorn to launch / manager Uvicorn workers ( uvicorn main:app--workers=2 )

I am not sure the performance difference between 2.1 and 2.2.
I am not sure the performance difference between "bringing up multiple workers in a container in a Pod" or "bringing up a worker in a container in a Pod + managing loads by bringing up k8s HPA"

For running one worker case like these, I think the these cases don't make much difference.
1. uvicorn worker -- _uvicorn main:app_
2. gunicorn ( running one uvicorn worker ) -- _gunicorn main:app -k uvicorn.workers.UvicornWorker_

For running multiple workers case, the question can be more which do better job in worker process management and if there are abt functionality you want to use in Gunicorn.

**Comment by mauermbq** [comment]
>For running multiple workers case, the question is about which one does better job in worker process management
for me this sounds that for K8S we do not need gunicorn. Scaling out many workers in one container is not the proposed way. Why not using inbuilt autoscaler? https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/

**Comment by Gatsby-Lee** [comment]
> > For running multiple workers case, the question is about which one does better job in worker process management
> > for me this sounds that for K8S we do not need gunicorn. Scaling out many workers in one container is not the proposed way. Why not using inbuilt autoscaler? https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/

you might want to use gunicorn if you run more than one Uvicorn worker in a container in a Pod.
It might be possible option. Yes. it is comparision b/w 2 Pod ( each Pod running a Uvicorn worker ) or 1 Pod ( running two Uvicorn worker ). 

You use HPA to bring up more Pods that run 2 Uvicorn workers in a Pod.
It's possible option depending on what we want to do :)

**Comment by elpablete** [comment]
Since there seems to be no better place for this discussion I would like to keep this going. I'm running fastapi on a cluster and I'm willing to do some testing to sort this out. The thing is, I don't really know how to compare the two cases:

1. One uvicorn worker per pod for two total pods (no gunicorn).
2. One pod with two uvicorn workers behind gunicorn. 

For example: is it fair to have two nodes for case 1? versus just one node for case 2?

I will share my results so we can all improve on fastapi + k8s.

Thanks in advance!

**Comment by fazlulkarimweb** [comment]
> Since there seems to be no better place for this discussion I would like to keep this going. I'm running fastapi on a cluster and I'm willing to do some testing to sort this out. The thing is, I don't really know how to compare the two cases:
> 
> 1. One uvicorn worker per pod for two total pods (no gunicorn).
> 2. One pod with two uvicorn workers behind gunicorn.
> 
> For example: is it fair to have two nodes for case 1? versus just one node for case 2?
> 
> I will share my results so we can all improve on fastapi + k8s.
> 
> Thanks in advance!

Would you please share your test with us? It will be helpful. Thanks in advance.

**Comment by elpablete** [comment]
I got carried away due to none related bugs. I still haven't done the tests. Will share when I actually manage to do them

**Comment by nreith** [comment]
I have come across this question in previous work and after lots of load testing and googling, I just applied my own peculiar logic. For me it doesn't matter as much if the asgi processes that are fast API and starlette are run by uvicorn only or unicorn with gunicorn managing workers, or be even nginx unit which supports asgi also. I have seen benchmarks showing that nginx unit blows uvicorn out of the water btw.

I want async, so asgi, not uwsgi. And I would like to maximize throughput while minimizing computer costs. For me it meant several workers per container, and also autoscaling pods. We simply tried a few combinations load testing with locust.io and simulating 2x and 3x peak traffic as well as seeing how long new pods take to spin up and how many requests went slow or failed during that time.

A helpfu

**Comment by GuiGav** [comment]
I'll add a fairly important point while running on Kubernetes and trying to monitor the application's health.
I have a RAM intensive application loading dataframes. The container memory limit can easily be reached and thus the process will be killed.

The difference is : 
- with gunicorn, gunicorn will kill the process and restart a new uvicorn worker. No kubernetes metrics will show the problem as the container did not restart
- without gunicorn, k8s will stop and restart the container, and the OOM and restarts can be monitored with metrics such as `kube_pod_container_status_last_terminated_reason` and `kube_pod_container_status_restarts_total`

One question that remains on my side is how gunicorn knows the memory limit and to kill the process before k8s does
